### PR TITLE
Created a time machine layout class for conversation messages cells.

### PR DESCRIPTION
**Pull request by MartinJibber**
There are still lots of layout improvements to make and some functionality to restore, but this PR gets most of the hard work done for the time machine layout.

### CODE REVIEW

File: iOS/Flows/Conversation/Cells/ConversationMessagesCell.swift [hunk 1]
```diff
@@ -34,18 +34,17 @@ class ConversationMessagesCell: UICollectionViewCell, ConversationMessageCellLay
     var conversation: MessageSequence?
 
     /// The maximum number of messages we'll show per stack of messages.
-    private let maxMessagesPerSection = 3
+    private let maxMessagesPerSection = 25
 
     override init(frame: CGRect) {
         super.init(frame: frame)
 
+        self.collectionView.decelerationRate = .fast
```
Comment (MartinJibber) [inline iOS/Flows/Conversation/Cells/ConversationMessagesCell.swift:42]: Making the paging feel a little snappier.

File: iOS/Flows/Conversation/Cells/ConversationMessagesCell.swift [hunk 2]
```diff
@@ -34,18 +34,17 @@ class ConversationMessagesCell: UICollectionViewCell, ConversationMessageCellLay
     var conversation: MessageSequence?
 
     /// The maximum number of messages we'll show per stack of messages.
-    private let maxMessagesPerSection = 3
+    private let maxMessagesPerSection = 25
```
Comment (MartinJibber) [inline iOS/Flows/Conversation/Cells/ConversationMessagesCell.swift:37]: Now that we can scroll, we should should a lot more messages.

File: iOS/Flows/Conversation/Cells/ConversationMessagesCell.swift [hunk 3]
```diff
@@ -20,7 +20,7 @@ class ConversationMessagesCell: UICollectionViewCell, ConversationMessageCellLay
     var handleTappedConversation: ((MessageSequence) -> Void)?
     var handleDeleteConversation: ((MessageSequence) -> Void)?
 
-    private lazy var collectionLayout = ConversationMessagesCellLayout(conversationDelegate: self)
+    private lazy var collectionLayout = TimelineCollectionViewLayout()
```
Comment (MartinJibber) [inline iOS/Flows/Conversation/Cells/ConversationMessagesCell.swift:23]: Switching to the new layout.

File: iOS/Flows/Conversation/Cells/ConversationMessagesCell.swift [hunk 4]
```diff
@@ -113,11 +110,12 @@ class ConversationMessagesCell: UICollectionViewCell, ConversationMessageCellLay
...
-        
-        UIView.animate(withDuration: 0.2) {
-            self.collectionLayout.showMessageStatus = isCentered
-        }
+        // TODO: Restore this
+//        guard self.collectionLayout.showMessageStatus != isCentered else { return }
+//
+//        UIView.animate(withDuration: 0.2) {
+//            self.collectionLayout.showMessageStatus = isCentered
+//        }
```
Comment (MartinJibber) [inline iOS/Flows/Conversation/Cells/ConversationMessagesCell.swift:118]: The message status labels haven't been implemented in the new layout yet.

File: iOS/Flows/Conversation/Cells/ConversationMessagesCell.swift [hunk 5]
```diff
@@ -132,13 +130,15 @@ class ConversationMessagesCell: UICollectionViewCell, ConversationMessageCellLay
...
-        }
+        // TODO: Restore this
+        return .zero
+//        if let frontCellIndex = self.collectionLayout
+//            .getFrontmostItemIndexPath(inSection: ConversationMessageSection.currentUserMessages.rawValue),
+//           let frontUserCell = self.collectionView.cellForItem(at: frontCellIndex) {
+//
+//            let overlayRect = frontUserCell.convert(frontUserCell.bounds, to: self)
+//            return self.convert(overlayRect, to: targetView)
+//        }
```
Comment (MartinJibber) [inline iOS/Flows/Conversation/Cells/ConversationMessagesCell.swift:141]: I need to update the code that determines the drop zone frame for the new layout.

File: iOS/Flows/Conversation/Cells/ConversationMessageCellLayoutAttributes.swift [hunk 6]
```diff
@@ -14,6 +14,8 @@ class ConversationMessageCellLayoutAttributes: UICollectionViewLayoutAttributes
     var shouldShowText: Bool = true
     /// The color of the background of the cell.
     var backgroundColor: Color = .lightGray
+    /// How bright the background color is. 0 is black. 1 is full brightness of the given color
+    var brightness: CGFloat = 1
```
Comment (MartinJibber) [inline iOS/Flows/Conversation/Cells/ConversationMessageCellLayoutAttributes.swift:18]: Using this variable to darken the messages the farther away they are.

File: iOS/Flows/Conversation/Cells/ConversationMessageCellDatasource.swift [hunk 7]
```diff
@@ -85,3 +85,16 @@ extension ConversationMessageCellDataSource {
...
+extension ConversationMessageCellDataSource: TimelineCollectionViewLayoutDataSource {
+
+    func getMessage(at indexPath: IndexPath) -> Messageable? {
+        guard let item = self.itemIdentifier(for: indexPath) else { return nil }
+        let messageController = ChatClient.shared.messageController(cid: item.channelID,
+                                                                    messageId: item.messageID)
+
+        return messageController.message
+    }
+}
```
Comment (MartinJibber) [inline iOS/Flows/Conversation/Cells/ConversationMessageCellDatasource.swift:100]: I made a data source delegate protocol for the time machine layout so it doesn't need an explicit reference to the specific datasource type.

File: iOS/Flows/Conversation/Cells/TimeMachineCollectionViewLayout.swift [hunk 8]
```diff
@@ -0,0 +1,304 @@
...
+    }
+
+    override func layoutAttributesForItem(at indexPath: IndexPath) -> UICollectionViewLayoutAttributes? {
+        // If the attributes are cached already, just return those.
+        if let attributes = self.cellLayoutAttributes[indexPath]  {
+            return attributes
+        }
+
+        // All items in a section are positioned relative to its frontmost item.
+        guard let frontmostIndexPath = self.getFrontmostIndexPath(in: indexPath.section) else { return nil }
```
Comment (MartinJibber) [inline iOS/Flows/Conversation/Cells/TimeMachineCollectionViewLayout.swift:143]: We don't want any gaps in the message stack so item positions are always relative to the frontmost item.

File: iOS/Flows/Conversation/Cells/TimeMachineCollectionViewLayout.swift [hunk 9]
```diff
@@ -0,0 +1,304 @@
...
+
+    /// A cache of item layout attributes so they don't have to be recalculated.
+    private var cellLayoutAttributes: [IndexPath : UICollectionViewLayoutAttributes] = [:]
+    /// A dictionary of z ranges for all the items. A z range represents the range that each item will be frontmost in its section
+    /// and its scale will be unaltered.
+    private var zRangesDict: [IndexPath : Range<CGFloat>] = [:]
+    /// The current position along the Z axis. This is based off of the collectionview's Y content offset.
+    private var zPosition: CGFloat {
+        return self.collectionView?.contentOffset.y ?? 0
+    }
```
Comment (MartinJibber) [inline iOS/Flows/Conversation/Cells/TimeMachineCollectionViewLayout.swift:35]: The items are laid out along the z axis. I'm using the y content offset to simulate moving along the z axis.

File: iOS/Flows/Conversation/Cells/TimeMachineCollectionViewLayout.swift [hunk 10]
```diff
@@ -0,0 +1,304 @@
...
+    // MARK: - UICollectionViewLayout Overrides
+
+    override var collectionViewContentSize: CGSize {
+        get {
+            guard let collectionView = collectionView, collectionView.frame != .zero else { return .zero }
+
+            let itemCount = CGFloat(self.numberOfItems(inSection: 0) + self.numberOfItems(inSection: 1))
+            var height = (itemCount - 1) * self.itemSize.height
+            height += collectionView.bounds.height
+            return CGSize(width: collectionView.bounds.width, height: height)
```
Comment (MartinJibber) [inline iOS/Flows/Conversation/Cells/TimeMachineCollectionViewLayout.swift:54]: Make the content offset big enough to scroll from the beginning end of the items.

File: iOS/Flows/Conversation/Cells/TimeMachineCollectionViewLayout.swift [hunk 11]
```diff
@@ -0,0 +1,304 @@
...
+                        endZ = CGFloat(nextIndex) * self.itemSize.height - self.itemSize.height
+                        break
+                    } else if nextIndex == sortedItemIndexPaths.count - 1 {
+                        endZ = CGFloat(nextIndex) * self.itemSize.height
+                        break
+                    }
+                }
+            }
+
+            self.zRangesDict[indexPath] = startZ..<endZ
```
Comment (MartinJibber) [inline iOS/Flows/Conversation/Cells/TimeMachineCollectionViewLayout.swift:123]: Each item has a range of z values where it is the frontmost in the stack. This function is calculating that.

File: iOS/Flows/Conversation/Cells/TimeMachineCollectionViewLayout.swift [hunk 12]
```diff
@@ -0,0 +1,304 @@
...
+            let normalized = (-vectorToCurrentZ)/self.itemSize.height
+            scale = clamp(normalized, max: 1) + 1
+            yOffset = normalized * -self.itemSize.height * 2
+            alpha = 1 - normalized
+        } else {
+            // If current z position is within the item's z range, don't adjust its scale or position.
+            scale = 1
+            yOffset = 0
+            alpha = 1
+        }
```
Comment (MartinJibber) [inline iOS/Flows/Conversation/Cells/TimeMachineCollectionViewLayout.swift:179]: Logic to figure out how much to grow or shrink the items.

**Review by Benjamindodgson — APPROVED** [review]
Love

**Pull request merged**